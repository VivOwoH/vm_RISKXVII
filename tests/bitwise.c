// XOR, OR, AND
int volatile *const ConsoleWriteSInt = (int *)0x0804;
char volatile *const ConsoleWriteChar = (char *)0x0800;

inline int scan_int() {
    int result;
    int addr = 0x0816; // read int
    asm volatile("lw %[res], 0(%[adr])" // load 
                    : [res]"=r"(result) // output
                    : [adr]"r"(addr)); // input
    return result;
}

int main() {
    int result[3]; 
    int count = 0;

    int mode = scan_int();

    int a = scan_int();
    int b = scan_int();
    result[0] = a ^ b; // XOR
    
    int c = scan_int();
    int d = scan_int();
    result[1] = c | d; // OR

    int e = scan_int();
    int f = scan_int();
    result[2] = e & f; // AND

    if (mode == 'X')
        *ConsoleWriteSInt = result[0];
    else if (mode == 'O')
        *ConsoleWriteSInt = result[1];
    else if (mode == 'A')
        *ConsoleWriteSInt = result[2];

    return 0;
}