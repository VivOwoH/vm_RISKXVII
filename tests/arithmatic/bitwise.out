8611832CPU Halt Requested
